examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,990 @@
+PetscErrorCode VecC2P(CeedVector x_ceed, PetscMemType mem_type, Vec X_petsc);
+PetscErrorCode VecReadP2C(Vec X_petsc, PetscMemType *mem_type, CeedVector x_ceed);
+PetscErrorCode VecReadC2P(CeedVector x_ceed, PetscMemType mem_type, Vec X_petsc);
+PetscErrorCode DMSwarmPICFieldP2C(DM dm_swarm, const char *field, CeedVector x_ceed);
+PetscErrorCode DMSwarmPICFieldC2P(DM dm_swarm, const char *field, CeedVector x_ceed);
+
+PetscScalar    EvalU_Poly(PetscInt dim, const PetscScalar x[]);
+PetscScalar    EvalU_Tanh(PetscInt dim, const PetscScalar x[]);
+PetscErrorCode EvalU_Poly_proj(PetscInt dim, PetscReal t, const PetscReal x[], PetscInt num_comp, PetscScalar *u, void *ctx);
+PetscErrorCode EvalU_Tanh_proj(PetscInt dim, PetscReal t, const PetscReal x[], PetscInt num_comp, PetscScalar *u, void *ctx);
+
+PetscErrorCode DMSwarmInitalizePointLocations(DM dm_swarm, PetscInt num_points, PetscBool set_gauss_swarm, PetscBool set_uniform_swarm);
+PetscErrorCode DMSwarmCreateReferenceCoordinates(DM dm_swarm, IS *is_points, Vec *ref_coords);
+PetscErrorCode DMSwarmInterpolateFromCellToSwarm_Petsc(DM dm_swarm, const char *field, Vec U_mesh);
+PetscErrorCode DMSwarmInterpolateFromCellToSwarm_Ceed(DM dm_swarm, const char *field, Vec U_mesh);
+PetscErrorCode DMSwarmCheckSwarmValues(DM dm_swarm, const char *field, PetscScalar tolerance, DMFunc TrueSolution);
+
+PetscErrorCode DMSwarmCreateProjectionRHS(DM dm_swarm, const char *field, Vec B_mesh);
+PetscErrorCode MatMult_SwarmMass(Mat A, Vec U_mesh, Vec V_mesh);
+
+// ------------------------------------------------------------------------------------------------
+// main driver
+// ------------------------------------------------------------------------------------------------
+int main(int argc, char **argv) {
+  MPI_Comm  comm;
+  char      ceed_resource[PETSC_MAX_PATH_LEN] = "/cpu/self";
+  PetscBool test_mode = PETSC_FALSE, set_uniform_swarm = PETSC_FALSE, set_gauss_swarm = PETSC_FALSE, view_petsc_swarm = PETSC_FALSE,
+            view_ceed_swarm = PETSC_FALSE, use_polynomial_target = PETSC_FALSE;
+  PetscInt           dim = 3, num_comp = 1, num_points = 200, mesh_order = 1, solution_order = 3, q_extra = 3;
+  PetscScalar        tolerance = 1E-3;
+  DM                 dm_mesh, dm_swarm;
+  Vec                U_mesh;
+  DMSwarmCeedContext swarm_ceed_context;
+
+  PetscCall(PetscInitialize(&argc, &argv, NULL, help));
+  comm = PETSC_COMM_WORLD;
+
+  // Read command line options
+  PetscOptionsBegin(comm, NULL, "libCEED example using PETSc with DMSwarm", NULL);
+
+  PetscCall(PetscOptionsBool("-test", "Testing mode (do not print unless error is large)", NULL, test_mode, &test_mode, NULL));
+  PetscCall(PetscOptionsBool("-gauss_swarm", "Use gauss points for coordinates in swarm", NULL, set_gauss_swarm, &set_gauss_swarm, NULL));
+  PetscCall(PetscOptionsBool("-uniform_swarm", "Use uniform coordinates in swarm", NULL, set_uniform_swarm, &set_uniform_swarm, NULL));
```

Comment:
In an ideal world this should be an enum, but I don't think it really matter for this example.

Reply:
Yeah. My main hangup is that I couldn't figure out what to call the other version of the points distribution? (Though we don't really use it much currently it is nice having the option to have a different number of points in each cell to see what that does to accuracy)

Reply:
That's fair, maybe like "path" or "semicircle"?